File: Partner-Api/src/main/java/work/slhaf/partner/api/factory/AgentRegisterFactory.java
package work.slhaf.partner.api.factory;

import cn.hutool.core.bean.BeanUtil;
import org.reflections.util.ClasspathHelper;
import work.slhaf.partner.api.entity.AgentContext;
import work.slhaf.partner.api.factory.capability.CapabilityCheckFactory;
import work.slhaf.partner.api.factory.capability.CapabilityInjectFactory;
import work.slhaf.partner.api.factory.capability.CapabilityRegisterFactory;
import work.slhaf.partner.api.factory.config.ConfigLoaderFactory;
import work.slhaf.partner.api.factory.context.AgentRegisterContext;
import work.slhaf.partner.api.factory.exception.ExternalModuleLoadFailedException;
import work.slhaf.partner.api.factory.exception.ExternalModulePathNotExistException;
import work.slhaf.partner.api.factory.module.ModuleCheckFactory;
import work.slhaf.partner.api.factory.module.ModulePreHookExecuteFactory;
import work.slhaf.partner.api.factory.module.ModuleProxyFactory;
import work.slhaf.partner.api.factory.module.ModuleRegisterFactory;

import java.io.File;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class AgentRegisterFactory {

    private static final List<URL> urls = new ArrayList<>();

    private AgentRegisterFactory() {
    }

    public static AgentContext launch(String packageName) {
        urls.addAll(packageNameToURL(packageName));
        AgentRegisterContext registerContext = new AgentRegisterContext(urls);
        //流程
        //0. 加载配置
        new ConfigLoaderFactory().execute(registerContext);
        //1. 注册并检查Capability
        new CapabilityRegisterFactory().execute(registerContext);
        new CapabilityCheckFactory().execute(registerContext);
        //2. 注册并检查Module
        new ModuleCheckFactory().execute(registerContext);
        new ModuleRegisterFactory().execute(registerContext);
        //3. 为module通过动态代理添加PostHook逻辑并进行实例化
        new ModuleProxyFactory().execute(registerContext);
        //. 先一步注入Capability,避免因前hook逻辑存在针对能力的引用而报错
        new CapabilityInjectFactory().execute(registerContext);
        //. 执行模块PreHook逻辑
        new ModulePreHookExecuteFactory().execute(registerContext);

        AgentContext agentContext = new AgentContext();
        BeanUtil.copyProperties(registerContext, agentContext);
        return agentContext;
    }

    /**
     * 添加可扫描包
     * @param packageName 指定的包名
     */
    public static void addScanPackage(String packageName) {
        urls.addAll(packageNameToURL(packageName));
    }

    /**
     * 添加外部模块目录
     * @param externalPackagePath 指定的外部模块目录路径
     */
    public static void addScanDir(String externalPackagePath) {
        File file = new File(externalPackagePath);
        if (!file.exists() || !file.isDirectory()) {
            throw new ExternalModulePathNotExistException("不存在的外部模块目录: " + externalPackagePath);
        }
        try {
            for (File f : file.listFiles()) {
                if (f.getName().endsWith(".jar")) {
                    urls.add(f.toURI().toURL());
                }
            }
        } catch (Exception e) {
            throw new ExternalModuleLoadFailedException("外部模块URL获取失败: " + externalPackagePath, e);
        }
    }

    private static List<URL> packageNameToURL(String packageName) {
        return ClasspathHelper.forPackage(packageName).stream().toList();
    }


}

File: Partner-Framework/src/main/java/work/slhaf/partner/framework/agent/factory/AgentRegisterFactory.kt
package work.slhaf.partner.framework.agent.factory

import org.reflections.util.ClasspathHelper
import work.slhaf.partner.framework.agent.exception.FactoryExecutionException
import work.slhaf.partner.framework.agent.factory.capability.CapabilityAnnotationValidatorFactory
import work.slhaf.partner.framework.agent.factory.capability.CapabilityInjectorFactory
import work.slhaf.partner.framework.agent.factory.capability.CapabilityRegisterFactory
import work.slhaf.partner.framework.agent.factory.component.ComponentAnnotationValidatorFactory
import work.slhaf.partner.framework.agent.factory.component.ComponentInitHookExecutorFactory
import work.slhaf.partner.framework.agent.factory.component.ComponentInjectorFactory
import work.slhaf.partner.framework.agent.factory.component.ComponentRegisterFactory
import work.slhaf.partner.framework.agent.factory.context.AgentRegisterContext
import work.slhaf.partner.framework.agent.factory.context.ShutdownHookCollectorFactory
import java.io.File
import java.net.URL

/**
 * Agent 注册总入口，按固定顺序串联各 Factory。
 *
 * 启动流程:
 * 1. 校验 Component 注解
 * 2. 注册 Component/Module
 * 3. 完成 Module 注入
 * 4. 校验 Capability 注解
 * 5. 注册 Capability 代理与路由
 * 6. 注入 Capability
 * 7. 执行 Init Hook
 * 8. 收集 Shutdown Hook
 */
object AgentRegisterFactory {
    private val urls: MutableList<URL> = mutableListOf()

    @JvmStatic
    fun launch(packageName: String) {
        urls.addAll(packageNameToURL(packageName))
        val registerContext = AgentRegisterContext(urls)
        // 1. 校验 Component 级别注解是否合规，避免注入到异常位置
        ComponentAnnotationValidatorFactory().execute(registerContext)
        // 2. 收集所有的 AgentComponent 实例
        ComponentRegisterFactory().execute(registerContext)
        // 3. 对模块与额外组件进行模块依赖注入
        ComponentInjectorFactory().execute(registerContext)
        // 4. 校验 Capability 注解与方法关系
        CapabilityAnnotationValidatorFactory().execute(registerContext)
        // 5. 根据 Capability 相关的扫描结果构造 Capability 实例
        CapabilityRegisterFactory().execute(registerContext)
        // 6. 将 Capability 实例注入至各个 AgentComponent 中
        CapabilityInjectorFactory().execute(registerContext)
        // 7. 执行模块 Init Hook 逻辑
        ComponentInitHookExecutorFactory().execute(registerContext)
        // 8. 校验并收集 Shutdown Hook 逻辑，并添加至 AgentContext 中
        ShutdownHookCollectorFactory().execute(registerContext)
    }

    @JvmStatic
    fun addScanPackage(packageName: String) {
        urls.addAll(packageNameToURL(packageName))
    }

    @JvmStatic
    fun addScanDir(externalPackagePath: String) {
        val file = File(externalPackagePath)
        if (!file.exists() || !file.isDirectory) {
            return
        }
        try {
            val files = file.listFiles() ?: return
            if (files.isEmpty()) {
                return
            }
            files.asSequence()
                .filter { it.name.endsWith(".jar") }
                .forEach { urls.add(it.toURI().toURL()) }
        } catch (e: Exception) {
            throw FactoryExecutionException(
                "Failed to load external module URLs from: $externalPackagePath",
                "agent-register-factory",
                e
            )
        }
    }

    private fun packageNameToURL(packageName: String): List<URL> {
        return ClasspathHelper.forPackage(packageName).toList()
    }
}
